compiler/src/dotty/tools/dotc/transform/HoistSuperArgs.scala

```diff
@@ -0,0 +1,171 @@
+package dotty.tools.dotc
+package transform
+
+import TreeTransforms._
+import core.DenotTransformers._
+import core.Symbols._
+import core.Contexts._
+import ast.TreeTypeMap
+import core.Types._
+import core.Flags._
+import core.Decorators._
+import collection.mutable
+import ast.Trees._
+import core.Names.TermName
+import core.NameKinds.SuperArgName
+import SymUtils._
+
+/** This phase hoists complex arguments to supercalls out of the enclosing class.
+ *  Example:
+ *
+ *      class B(y: Int) extends A({ def f(x: Int) = x * x; f(y)})
+ *
+ *  is translated to
+ *
+ *      class B(y: Int) extends A(B#B$superArg$1(this.y)) {
+ *        private <static> def B$superArg$1(y: Int): Int = {
+ *          def f(x: Int): Int = x.*(x); f(y)
+ *        }
+ *      }
+ *
+ *  An argument is complex if it contains a method or template definition, a this or a new,
+ *  or it contains an identifier which needs a `this` prefix to be accessed. This is the case
+ *  if the identifer neither a global reference nor a reference to a parameter of the enclosing class.
+ *  @see needsHoist for an implementation.
+ *
+ *  A hoisted argument definition gets the parameters of the class it is hoisted from
+ *  as method parameters. The definition is installed in the scope enclosing the class,
+ *  or, if that is a package, it is made a static method of the class itself.
+ */
+class HoistSuperArgs extends MiniPhaseTransform with IdentityDenotTransformer { thisTransform =>
+  import ast.tpd._
+
+  def phaseName = "hoistSuperArgs"
+
+  /** Hoist complex arguments in super call `parent` out of the class.
+   *  @return A pair consisting of the transformed super call and a list of super argument
+   *          defininitions.
+   */
+  def hoistSuperArgs(parent: Tree, cls: Symbol)(implicit ctx: Context): (Tree, List[DefDef]) = {
+    lazy val constr = cls.primaryConstructor
+    lazy val allParams = // The parameters that can be accessed in the supercall
+      cls.info.decls.filter(d => d.is(TypeParam) || d.is(TermParamAccessor))
+    val superArgDefs = new mutable.ListBuffer[DefDef]
+
+    /** The parameter references defined by the primary constructor info */
+    def allParamRefs(tp: Type): List[ParamRef] = tp match {
+      case tp: LambdaType => tp.paramRefs ++ allParamRefs(tp.resultType)
+      case _ => Nil
+    }
+
+    /** Splice `restpe` in final result type position of `tp` */
+    def replaceResult(tp: Type, restpe: Type): Type = tp match {
+      case tp: LambdaType =>
+        tp.derivedLambdaType(resType = replaceResult(tp.resultType, restpe))
+      case _ => restpe
+    }
+
+    /** A method representing a hoisted supercall argument */
+    def newSuperArgMethod(argType: Type) = {
+      val (staticFlag, methOwner) =
+        if (cls.owner.is(Package)) (JavaStatic, cls) else (EmptyFlags, cls.owner)
+      val argTypeWrtConstr = argType.subst(allParams, allParamRefs(constr.info))
```

Comment:
It seems `allParamRefs(constr.info)` is called for each argument, maybe turn it into lazy val? Or, it's because `constr.info` may change between different calls?